module/zfs/bpobj.c

```diff
@@ -242,119 +277,171 @@ bpobj_iterate_impl(bpobj_t *bpo, bpobj_itor_t func, void *arg, dmu_tx_t *tx,
 			if (bpo->bpo_havecomp) {
-				bpo->bpo_phys->bpo_comp -= BP_GET_PSIZE(bp);
-				bpo->bpo_phys->bpo_uncomp -= BP_GET_UCSIZE(bp);
+				comp_freed += BP_GET_PSIZE(bp);
+				uncomp_freed += BP_GET_UCSIZE(bp);
 			}
 			bpo->bpo_phys->bpo_num_blkptrs--;
 			ASSERT3S(bpo->bpo_phys->bpo_num_blkptrs, >=, 0);
 		}
 	}
+	if (free)
+		propagate_space_reduction(bpi, freed, comp_freed,
+		    uncomp_freed, tx);
 	if (dbuf) {
 		dmu_buf_rele(dbuf, FTAG);
 		dbuf = NULL;
 	}
 	if (free) {
-		VERIFY3U(0, ==, dmu_free_range(bpo->bpo_os, bpo->bpo_object,
-		    (i + 1) * sizeof (blkptr_t), DMU_OBJECT_END, tx));
+		VERIFY3U(0, ==, dmu_free_range(bpo->bpo_os,
+		    bpo->bpo_object, (i + 1) * sizeof (blkptr_t),
+		    DMU_OBJECT_END, tx));
 	}
-	if (err || !bpo->bpo_havesubobj || bpo->bpo_phys->bpo_subobjs == 0)
-		goto out;
+	return (err);
+}
 
-	ASSERT(bpo->bpo_havecomp);
-	err = dmu_object_info(bpo->bpo_os, bpo->bpo_phys->bpo_subobjs, &doi);
-	if (err) {
-		mutex_exit(&bpo->bpo_lock);
-		return (err);
+static void
+bpobj_iterate_cleanup(list_t *stack, int err)
+{
+	bpobj_t *bpo;
+
+	for (bpobj_info_t *bpi_cleanup = list_head(stack); bpi_cleanup != NULL;
+	    bpi_cleanup = list_next(stack, bpi_cleanup)) {
+		ASSERT3U(err, !=, 0);
+		bpo = bpi_cleanup->bpi_bpo;
+		ASSERT3P(bpo, !=, NULL);
+		if (bpi_cleanup->bpi_parent != NULL) {
+			bpobj_close(bpo);
+			kmem_free(bpi_cleanup->bpi_bpo, sizeof (bpobj_t));
+		}
+		kmem_free(bpi_cleanup, sizeof (bpobj_info_t));
 	}
-	ASSERT3U(doi.doi_type, ==, DMU_OT_BPOBJ_SUBOBJ);
-	epb = doi.doi_data_block_size / sizeof (uint64_t);
-
-	for (i = bpo->bpo_phys->bpo_num_subobjs - 1; i >= 0; i--) {
-		uint64_t *objarray;
-		uint64_t offset, blkoff;
-		bpobj_t sublist;
-		uint64_t used_before, comp_before, uncomp_before;
-		uint64_t used_after, comp_after, uncomp_after;
-
-		offset = i * sizeof (uint64_t);
-		blkoff = P2PHASE(i, epb);
+}
 
-		if (dbuf == NULL || dbuf->db_offset > offset) {
-			if (dbuf)
-				dmu_buf_rele(dbuf, FTAG);
-			err = dmu_buf_hold(bpo->bpo_os,
-			    bpo->bpo_phys->bpo_subobjs, offset, FTAG, &dbuf, 0);
-			if (err)
-				break;
-		}
+#define	SUBOBJS(x) x->bpo_phys->bpo_num_subobjs
 
-		ASSERT3U(offset, >=, dbuf->db_offset);
-		ASSERT3U(offset, <, dbuf->db_offset + dbuf->db_size);
+static int
+bpobj_iterate_impl(bpobj_t *initial_bpo, bpobj_itor_t func, void *arg,
+    dmu_tx_t *tx, boolean_t free)
+{
+	list_t stack;
+	bpobj_t *bpo = initial_bpo;
+	int err = 0;
 
-		objarray = dbuf->db_data;
-		err = bpobj_open(&sublist, bpo->bpo_os, objarray[blkoff]);
-		if (err)
-			break;
-		if (free) {
-			err = bpobj_space(&sublist,
-			    &used_before, &comp_before, &uncomp_before);
-			if (err != 0) {
-				bpobj_close(&sublist);
-				break;
+	list_create(&stack, sizeof (bpobj_info_t),
+	    offsetof(bpobj_info_t, bpi_node));
+	for (bpobj_info_t *bpi = bpi_alloc(bpo, NULL, 0, 0);
+	    bpi != NULL; bpi = list_remove_head(&stack)) {
+		bpo = bpi->bpi_bpo;
+		mutex_enter(&bpo->bpo_lock);
```

Comment:
`bpo_lock` is not exited in some of the error paths below - the ones where we `break`.